ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -313,3 +378,105 @@ prop_appendOldStateRestoresPreWindow TestChainState{..} =
 
 size :: Num b => Seq.StrictSeq a -> b
 size = fromIntegral . Seq.length
+
+data Input c
+  = InputEBB !(WithOrigin SlotNo) !SlotNo
+    -- ^ the preceding signed slot, the EBB's slot
+  | InputSigner !(CS.PBftSigner c)
+  deriving (Eq, Show)
+
+newtype Inputs c = Inputs {unInputs :: [Input c]}
+  deriving (Eq, Show)
+
+signatureInputs :: Inputs c -> [CS.PBftSigner c]
+signatureInputs (Inputs inps) = [ signer | InputSigner signer <- inps ]
+
+ebbInputs :: Inputs c -> [(WithOrigin SlotNo, SlotNo)]
+ebbInputs (Inputs inps) = [ (prev, slot) | InputEBB prev slot <- inps ]
+
+slotInput :: Input c -> SlotNo
+slotInput = \case
+  InputEBB _ slot -> slot
+  InputSigner   x -> CS.pbftSignerSlotNo x
+
+slotInputs :: Inputs c -> [SlotNo]
+slotInputs = map slotInput . unInputs
+
+lengthInputs :: Num a => Inputs c -> a
+lengthInputs = fromIntegral . length . unInputs
+
+headSigner :: Inputs c -> CS.PBftSigner c
+headSigner = head . signatureInputs
```

Comment:
If you want, you can use `NonEmpty` list to avoid this partial `head`, but since this is test code, I didn't bother.